core/distributed/matrix_kernels.hpp

```diff
@@ -0,0 +1,86 @@
+/*******************************<GINKGO LICENSE>******************************
+Copyright (c) 2017-2022, the Ginkgo authors
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+
+1. Redistributions of source code must retain the above copyright
+notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_CORE_DISTRIBUTED_MATRIX_KERNELS_HPP_
+#define GKO_CORE_DISTRIBUTED_MATRIX_KERNELS_HPP_
+
+
+#include <ginkgo/core/base/array.hpp>
+#include <ginkgo/core/base/device_matrix_data.hpp>
+#include <ginkgo/core/base/matrix_data.hpp>
+#include <ginkgo/core/base/types.hpp>
+#include <ginkgo/core/distributed/partition.hpp>
+
+
+#include "core/base/kernel_declaration.hpp"
+
+
+namespace gko {
+namespace kernels {
+
+
+#define GKO_DECLARE_BUILD_DIAG_OFFDIAG(ValueType, LocalIndexType,      \
+                                       GlobalIndexType)                \
+    void build_diag_offdiag(                                           \
+        std::shared_ptr<const DefaultExecutor> exec,                   \
+        const device_matrix_data<ValueType, GlobalIndexType>& input,   \
+        const distributed::Partition<LocalIndexType, GlobalIndexType>* \
+            row_partition,                                             \
+        const distributed::Partition<LocalIndexType, GlobalIndexType>* \
+            col_partition,                                             \
+        comm_index_type local_part,                                    \
+        device_matrix_data<ValueType, LocalIndexType>& diag_data,      \
+        device_matrix_data<ValueType, LocalIndexType>& offdiag_data,   \
```

Comment:
I need to set the size/dim of the data, which at least for the offdiag case is only known during the kernel.

Reply:
good point. I am really not happy with mixing `core` and kernel code in that fashion, an alternative would be a single out-parameter for the number of offdiag columns?

Reply:
So I guess the only core code that is acceptable is array code, right? I would still need either the resize or the assign operator.

Reply:
yes. At some point in the future (probably 2.0), we will need to split everything into a core-core (probably executor-only stuff + header-only types like Array that depend directly on it) and regular core (all other Ginkgo types) so that core and kernels depend on core-core, and core depends on kernels.

Reply:
perhaps to clarify what I'm doing, I create a new device matrix data with the correct size and move that into the input data. So the only implementation that I need here is the device_matrix_data constructor that takes the arrays as parameter. Do you expect that we also move that constructor into core?

Reply:
hmm, that's difficult to judge. I would prefer if we kept them separate (especially because in Windows, all members of a class need to be part of the same DLL unless you tag each one individually with different __dllspec attributes, which misses some special member functions etc.), but that would mean that the constructor can't be called from the kernel.